### PR TITLE
Various fixes and code refactoring

### PR DESCRIPTION
**Pull request by bvanassche**
Hello Jens,

As you probably know I'm still working on adding zoned block device support to fio. Although I hope to finish that work soon I already want to send you a pull request with various changes I came up with while working on zoned block device support. Please consider these patches for the official fio repository.

Thanks,

Bart.

### CODE REVIEW

**Comment by axboe** [comment]
Looks good to me, thanks Bart.

File: parse.h [hunk 1]
```diff
@@ -117,7 +117,7 @@ static inline void *td_var(void *to, struct fio_option *o, unsigned int offset)
 	else
 		ret = to;
 
-	return (char *) ret + offset;
+	return ret + offset;
```
Comment (sitsofe) [inline parse.h]: Void pointer arithmetic is technically illegal in C (https://stackoverflow.com/a/3524270/9109338 ) and is a GCC extension...

File: parse.c [hunk 2]
```diff
@@ -71,10 +71,11 @@ static void show_option_range(const struct fio_option *o,
 			      size_t (*logger)(const char *format, ...))
 {
 	if (o->type == FIO_OPT_FLOAT_LIST) {
-		if (o->minfp == DBL_MIN && o->maxfp == DBL_MAX)
+		if (!o->minfp && !o->maxfp)
 			return;
 
-		logger("%20s: min=%f", "range", o->minfp);
+		if (o->minfp != DBL_MIN)
+			logger("%20s: min=%f", "range", o->minfp);
```
Comment (sitsofe) [inline parse.c]: If you were to only set o->maxfp the logger output might look strange

File: parse.c [hunk 3]
```diff
@@ -668,15 +669,17 @@ static int __handle_option(const struct fio_option *o, const char *ptr,
...
-		if (uf > o->maxfp) {
-			log_err("value out of range: %f"
-				" (range max: %f)\n", uf, o->maxfp);
-			return 1;
-		}
-		if (uf < o->minfp) {
-			log_err("value out of range: %f"
-				" (range min: %f)\n", uf, o->minfp);
-			return 1;
+		if (o->minfp || o->maxfp) {
```
Comment (sitsofe) [inline parse.c]: So if either of the fp limits is set you check both of them - is this correct or should the guard be rolled into the sub conditions?

File: parse.c (hunk 2 above)
Comment (bvanassche) [reply parse.c]: Can you clarify your comment? Either both minfp and maxfp should be set or none of them so I don't see the issue.

File: parse.c (hunk 3 above)
Comment (bvanassche) [reply parse.c]: Yes, if any of the fp limits has been set both are checked. So if someone only wants to set one limit the other should be set to DBL_MIN or DBL_MAX as appropriate.

File: parse.c (hunk 2 above)
Comment (sitsofe) [reply parse.c]: Imagine I want a range from 0 - DBL_MAX then won't I only see ", max=..." in the logs?

File: parse.c (hunk 3 above)
Comment (sitsofe) [reply parse.c]: OK that makes sense.

File: parse.h (hunk 1 above)
Comment (axboe) [reply parse.h]: Yeah I do know that, but it's supported everywhere. I did ponder dropping this patch, but I'd be very surprised if I don't already do pointer math in a lot of other places in fio.

Comment (axboe) [reply parse.h]: Adding -Wpointer-arith gives me 103 warnings :-)

File: parse.c (hunk 2 above)
Comment (bvanassche) [reply parse.c]: With the following patch:

diff --git a/options.c b/options.c
index 17d7245f5490..3cd84bfe0608 100644
--- a/options.c
+++ b/options.c
@@ -3504,7 +3511,7 @@ struct fio_option fio_options[FIO_MAX_OPTS] = {
 		.def	= "100",
 		.maxlen	= 1,
 		.minfp	= 0.0,
-		.maxfp	= 100.0,
+		.maxfp	= DBL_MAX,
 		.category = FIO_OPT_C_IO,
 		.group	= FIO_OPT_G_LATPROF,
 	},

fio shows the following output:

./fio --cmdhelp=latency_percentile
  latency_percentile: Percentile of IOs must be below latency_target
                type: list of floating point values separated by ':' (opt=5.9:7.8)
             default: 100
               range: min=0.000000

I think that's the output we all expect?

Comment (sitsofe) [reply parse.c]: I had it the wrong way round - o->minfp would have to be DBL_MIN and o->maxfp would have to be something other than DBL_MAX:

patch:
```patch
diff --git a/options.c b/options.c
index 17d7245f..0b7c3076 100644
--- a/options.c
+++ b/options.c
@@ -6,6 +6,7 @@
 #include <assert.h>
 #include <sys/stat.h>
 #include <netinet/in.h>
+#include <float.h>
 
 #include "fio.h"
 #include "verify.h"
@@ -3503,7 +3504,7 @@ struct fio_option fio_options[FIO_MAX_OPTS] = {
                .help   = "Percentile of IOs must be below latency_target",
                .def    = "100",
                .maxlen = 1,
-               .minfp  = 0.0,
+               .minfp  = DBL_MIN,
                .maxfp  = 100.0,
                .category = FIO_OPT_C_IO,
                .group  = FIO_OPT_G_LATPROF,
```
example:
```
$ ./fio --cmdhelp=latency_percentile
  latency_percentile: Percentile of IOs must be below latency_target
                type: list of floating point values separated by ':' (opt=5.9:7.8)
             default: 100
, max=100.000000
```

Comment (bvanassche) [reply parse.c]: See also https://github.com/axboe/fio/pull/574

Comment (sitsofe) [reply parse.c]: Yup that got it.

**Pull request merged**